src/lib/Ouroboros/Byron/Proxy/Main.hs

```diff
@@ -133,42 +142,178 @@ data ByronProxy = ByronProxy
+    , handleReq = \txId -> atomically $ do
+        snapshot <- Tx.Out.mempoolGetSnapshot reader
+        ticketMap <- readTVar txTicketMapVar
+        let outcome = do
+              ticketNo <- Map.lookup (untag txId) ticketMap
+              Tx.Out.mempoolLookupTx snapshot ticketNo
+        pure $ fmap toByronTxMsgContents outcome
+    -- Peer gave us this transaction. Put it into the pool and give False to
+    -- mean it should _not_ be relayed.
+    , handleData = \txMsgContents -> do
+        let tx   = fromByronTxMsgContents txMsgContents
+            txid = Tx.In.txId writer tx
+        outcome <- Tx.In.mempoolAddTxs writer [tx]
+        case outcome of
+          []          -> pure False
+          -- Relay if it was added, and remember the ticket number of the
+          -- transaction id.
+          (txid' : _) -> pure (txid == txid')
+    }
+  where
+    ptag :: Proxy TxMsgContents
+    ptag = Proxy
+    keyFromValue :: TxMsgContents -> TxId
+    keyFromValue = hash . taTx . getTxMsgContents
+    reader :: Tx.Out.TxSubmissionMempoolReader (GenTxId (Block cfg)) (GenTx (Block cfg)) TicketNo IO
+    reader = getMempoolReader mempool
+    writer :: Tx.In.TxSubmissionMempoolWriter (GenTxId (Block cfg)) (GenTx (Block cfg)) TicketNo IO
+    writer = getMempoolWriter mempool
+
+-- | Try to send every transaction found in the mempool out to the Byron
+-- network.
+--
+-- It's possible that some transactions are missed, perhaps because `sendTx`
+-- takes too long a time. But that's ok.
+-- It's also possible that the transactions sent by this were actually received
+-- from Byron, and relayed by the Byron inv/req/data system. That's also ok:
+-- if a peer already has it, it won't ask for it (it will ignore the inv).
+sendTxsFromMempool :: Mempool IO (Block cfg) TicketNo -> Diffusion IO -> IO x
+sendTxsFromMempool mempool diff = go (Mempool.zeroIdx mempool)
+  where
+    go ticketNo = do
+      -- Block until a non-empty list of transactions to send appears.
+      txsToSend <- atomically $ do
+        snapshot <- Mempool.getSnapshot mempool
+        let txsToSend = Mempool.snapshotTxsAfter snapshot ticketNo
+        check (not (null txsToSend))
+        pure txsToSend
+      -- Left fold to take the TicketNo of the rightmost member.
+      -- Mempool claims this will always be increasing.
+      ticketNo' <- foldlM folder ticketNo txsToSend
+      go ticketNo'
+
+    -- Left fold action over a list of transactions to send: send each one and
+    -- keep the highest ticket number.
+    folder = \_ (tx, idx) -> do
+      -- Result is a Bool indicating whether it was accepted.
+      -- We don't care.
+      _ <- sendTx diff (toByronTxAux tx)
+      pure idx
+
+-- | Keep a map from Byron transaction ids to ticket numbers in sync with the
+-- mempool. Of course there could be some lag depending on the scheduler.
+updateMempoolIdMap
+  :: forall cfg x .
+     Mempool IO (Block cfg) TicketNo
+  -> TVar (Map TxId TicketNo)
+  -> IO x
```

Comment:
I've been using `Void` to indicate that one never constructs the returned value, it stands out a bit more than `forall x. ... -> IO x`.

Reply:
Changed to `forall void. ... -> IO void`. This way you can use it anywhere without importing `Data.Void` and using `fmap absurd`.